### PR TITLE
fix[tool]: update VarAccess pickle implementation

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
fix a bug where unpickling `annotated_vyper_module` would lead to a crash:
```
AttributeError: 'VarAccess' object has no attribute 'variable'
```
a repro can be found at https://github.com/charles-cooper/vyper/commit/2a5facf43ffb08b66f76d3d2b16b8f72f22d0625.

this is a blocker for tooling, for instance titanoboa relies on pickle/unpickle to cache CompilerData objects:
https://github.com/vyperlang/titanoboa/blob/86df8936654db20686410488738d7abaf165a4c9/boa/util/disk_cache.py#L65-L66

the underlying issue is that `pickle.loads()` calls `obj.__hash__()` for objects that are keys in a hashed data structure - namely, dicts, sets and frozensets. this causes a crash when there is a cycle in the object graph, because the object is not fully instantiated at the time that `__hash__()` is called. this is a cpython bug, reported at https://github.com/python/cpython/issues/124937.

@serhiy-storchaka suggested the approach taken in this PR, which breaks the loop before pickling: https://github.com/python/cpython/issues/124937#issuecomment-2392227290.

### Commit message

````
fix a bug where unpickling `annotated_vyper_module` would lead to
a crash:

```
AttributeError: 'VarAccess' object has no attribute 'variable'
```

this is a blocker for tooling, for instance, titanoboa
relies on pickle/unpickle to cache `CompilerData` objects:
https://github.com/vyperlang/titanoboa/blob/86df8936654db2068641/boa/util/disk_cache.py#L65-L66

the underlying issue is that `pickle.loads()` calls `obj.__hash__()`
for objects that are keys in a hashed data structure - namely, dicts,
sets and frozensets. this causes a crash when there is a cycle in
the object graph, because the object is not fully instantiated at the
time that `__hash__()` is called. this is a cpython issue, reported at
python/cpython#124937.

@serhiy-storchaka suggested the approach taken in this PR, which breaks
the loop before pickling:
https://github.com/python/cpython/issues/124937#issuecomment-2392227290

note that the implementation of `__reduce__()` in this PR is safe, since
there is no cycle in the hash function itself, since the recursion
breaks in `VarInfo.__hash__()`. in other words, there is no possibility
of `VarAccess.__hash__()` changing mid-way through reconstructing the
object.
````
### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: vyper/semantics/analysis/base.py [hunk 1]
```diff
@@ -234,6 +234,18 @@ class VarAccess:
...
     SUBSCRIPT_ACCESS: ClassVar[Any] = object()
 
+    def __reduce__(self):
+        dict_obj = {f.name: getattr(self, f.name) for f in fields(self)}
+        return self.__class__._produce, (dict_obj,)
+
+    @classmethod
+    def _produce(cls, data):
+        ret = cls.__new__(cls)
+        # bypass blocking of setattr by `frozen=True`
```
Comment (cyberthirst) [inline vyper/semantics/analysis/base.py:244]: should we use `frozen` instances if we use this bypass?

Comment (charles-cooper) [reply vyper/semantics/analysis/base.py:244]: Yes, this is only during a specialized instance of object construction to work around a bug in pickle. The fields cannot be modified after construction.

Comment (charles-cooper) [reply vyper/semantics/analysis/base.py:244]: See for instance how the default dataclasses setstate works:
https://github.com/python/cpython/blob/16cd6cc86b3ba20074ae3eefb61aeb24ee1544f7/Lib/dataclasses.py#L1198

**Comment by cyberthirst** [comment]
can we add an explainer comment on why's this necessary?

**Comment by charles-cooper** [comment]
> can we add an explainer comment on why's this necessary?

will get to it

File: vyper/semantics/analysis/base.py (hunk 1 above)
Comment (serhiy-storchaka) [reply vyper/semantics/analysis/base.py:247]: Why not simply return `cls(**data)`?

File: vyper/semantics/analysis/base.py [hunk 2]
```diff
@@ -234,6 +234,21 @@ class VarAccess:
     # A sentinel indicating a subscript access
     SUBSCRIPT_ACCESS: ClassVar[Any] = object()
 
+    # custom __reduce__ and _produce implementations to work around
+    # a pickle bug.
+    # see https://github.com/python/cpython/issues/124937#issuecomment-2392227290
+    def __reduce__(self):
+        dict_obj = {f.name: getattr(self, f.name) for f in fields(self)}
+        return self.__class__._produce, (dict_obj,)
```
Comment (serhiy-storchaka) [inline vyper/semantics/analysis/base.py:242]: You could also use the following trick if all fields can be passed as keyword arguments to constructor in subclasses:
```suggestion
        return partial(self.__class__, **dict_obj), ()
```

File: vyper/semantics/analysis/base.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/semantics/analysis/base.py]: good idea, thanks!
d3309e75ca12500d979bbc15f22808bea7f52c12

File: vyper/semantics/analysis/base.py (hunk 2 above)
Comment (charles-cooper) [reply vyper/semantics/analysis/base.py:242]: interesting, is there a performance difference?

**Comment by charles-cooper** [comment]
copying in some offline discussion with @cyberthirst about why the hash is safe even though the data structure is cyclic:

> There is no cycle in the hash function, it breaks at VarInfo.
> 
> https://github.com/vyperlang/vyper/blob/c02d2d8c5ed8c904c31b7f3f937ab01781fc9891/vyper/semantics/analysis/base.py#L182
> 
> So the contents of the _expr_info._reads/_writes can't affect the value of VarAccess hash.

File: vyper/semantics/analysis/base.py (hunk 2 above)
Comment (serhiy-storchaka) [reply vyper/semantics/analysis/base.py:242]: It may be slower. The advantage is that you do not need to add private method `_produce()`.

**Review by cyberthirst — APPROVED** [review]

**Pull request merged**